### PR TITLE
zeroize Key and derived keys upon drop

### PR DESCRIPTION
**Pull request by warner**
refs #63

@tarcieri hey, am I using this right?

### CODE REVIEW

File: src/core/receive.rs [hunk 1]
```diff
@@ -104,8 +106,10 @@ impl ReceiveMachine {
         phase: &Phase,
         body: &[u8],
     ) -> Option<Vec<u8>> {
-        let data_key = key::derive_phase_key(&side, &key, &phase);
-        key::decrypt_data(&data_key, body)
+        let mut data_key = key::derive_phase_key(&side, &key, &phase);
```
Comment (tarcieri) [inline src/core/receive.rs]: You can use `Zeroizing` to do zeroize-on-drop automatically: https://docs.rs/zeroize/latest/zeroize/struct.Zeroizing.html

```suggestion
        let mut data_key = Zeroizing::new(key::derive_phase_key(&side, &key, &phase));
```

File: src/core/send.rs [hunk 2]
```diff
@@ -42,10 +43,11 @@ impl SendMachine {
                 match event {
                     GotVerifiedKey(ref key) => {
                         for (phase, plaintext) in self.queue.drain(..) {
-                            let data_key =
+                            let mut data_key =
```
Comment (tarcieri) [inline src/core/send.rs]: Can use `Zeroizing` here (and below) as well

File: src/core/receive.rs (hunk 1 above)
Comment (warner) [reply src/core/receive.rs]: Cool, could I change the signature of `derive_phase_key` to do that automatically for everything it returns? It currently returns a `Vec<u8>`.

Comment (warner) [reply src/core/receive.rs]: what would the new signature be? `-> Zeroizing<Vec<u8>>` ?

Comment (tarcieri) [reply src/core/receive.rs]: Yup

Comment (warner) [reply src/core/receive.rs]: ok, dumb question, what's the cleanest way to convert the unit test at line 280-ish, from `assert_eq!(hex::encode(dk11), "abcabc..")`? That `dk11` was a `Vec<u8>`, now it's a `Zeroizing<Vec<u8>>`. The compiler is complaining that `std::convert::AsRef<[u8]>` is not implemented (`hex::encode` is `pub fn encode<T: AsRef<[u8]>>(data: T) -> String`).

Comment (warner) [reply src/core/receive.rs]: huh, `hex::decode(&*dk11)` worked, but surely that's silly?

Comment (tarcieri) [reply src/core/receive.rs]: It's doing deref coercion

Comment (warner) [reply src/core/receive.rs]: ok, so not silly (or at least not too silly to use). thanks!

**Pull request merged**